Add parameter feature to serialization and use it for CAddress

sipa opened the pull request:
This introduces a concept of "serialization parameters". Every serializer can define its own parameter type, and a value of that type must be provided when serializing/deserializing, e.g. using `ss << WithParams(parameter, value)`, causing `value` to be serialized to `ss`, with parameter `parameter`. Note these parameters are automatically passed down the stack, so for example a vector can be serialized with a parameter, which will be available to all the elements' serializers.

In this PR, the feature is only used for `CAddress` serialization, but I plan to also convert `SERIALIZE_TRANSACTION_NO_WITNESS` in a follow-up. Eventually I hope to get rid of all GetVersion/GetType and replace it with this approach, which is type-safe (code won't compile if the correct parameter isn't passed somewhere), and avoids collisions between flags.

sipa committed:
Disentangle disk address version from client version

sipa committed:
Support for serialization parameters

sipa committed:
Use serialization parameters for CAddress serialization

sipa committed:
Replace READWRITEAS macro with AsBase wrapping function

practicalswift commented:
Concept ACK

Getting rid of `GetVersion`/`GetType` and replacing them with a compile-time checked approach would be very nice.

jnewbery commented on src/protocol.h:
```diff
@@ -354,28 +354,43 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
+
+    // The only time we serialize a CAddress object without nTime is in
+    // the initial VERSION messages which contain two CAddress records.
+    // At that point, the serialization version is INIT_PROTO_VERSION.
```
This sentence can be removed now. It's meaningless without a serialization version.

sipa replied:
Arguably there still is a serialization version, but the link between that and CAddress serialization is purely historical. Removed by incorporating @MarcoFalke's changes below.

jnewbery reviewed (commented):
Concept ACK. Can be rebased on master now that #19486 is merged to drop the first two commits.

MarcoFalke commented on src/protocol.h:
```diff
@@ -368,23 +380,15 @@ class CAddress : public CService
-            }
+        } else if (fmt == AddrSerialization::NETWORK_WITHTIME) {
+            READWRITE(obj.nTime);
         }
```
in commit "user serialization ...":


Still not compatible. Now I get:

```
2020-07-13T17:01:39Z [opencon] trying connection 95.183.48.89:18333 lastseen=2468.3hrs
2020-07-13T17:01:39Z [opencon] Added connection peer=2
2020-07-13T17:01:39Z [opencon] sending version (103 bytes) peer=2
2020-07-13T17:01:39Z [opencon] send version message: version 70015, blocks=1746969, us=[::]:0, peer=2
2020-07-13T17:01:39Z [msghand] received: version (102 bytes) peer=2
2020-07-13T17:01:39Z [msghand] ProcessMessages(version, 102 bytes): Exception 'CDataStream::read(): end of data: iostream error' (NSt8ios_base7failureB5cxx11E) caught
2020-07-13T17:01:39Z [msghand] received: verack (0 bytes) peer=2
2020-07-13T17:01:39Z [msghand] Misbehaving: 95.183.48.89:18333 peer=2 (0 -> 1)

sipa replied:
There was an unrelated bug to the previous one in this commit, with the same effect: `addrFrom` in the VERSION message was parsed with `NETWORK_WITHTIME`.

MarcoFalke commented on src/protocol.h:
```diff
@@ -354,28 +354,43 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
+    // After the version handshake, all ADDR messages are serialized with
+    // nTime.
+    NETWORK_NOTIME,
+    NETWORK_WITHTIME
```
in the last commit:

Could add a trailing semicolon? I've prepared a suggested diff for you that also switches to a switch-case so that the compiler can check all cases are covered, feel free to take or reject:


```diff
diff --git a/src/protocol.h b/src/protocol.h
index 500cf346e9..0f5786a1a9 100644
--- a/src/protocol.h
+++ b/src/protocol.h
@@ -356,14 +356,8 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
 
 enum class AddrSerialization {
     DISK,
-
-    // The only time we serialize a CAddress object without nTime is in
-    // the initial VERSION messages which contain two CAddress records.
-    // At that point, the serialization version is INIT_PROTO_VERSION.
-    // After the version handshake, all ADDR messages are serialized with
-    // nTime.
     NETWORK_NOTIME,
-    NETWORK_WITHTIME
+    NETWORK_WITHTIME,
 };
 
 /** A CService with information about it as peer */
@@ -383,13 +377,23 @@ public:
     SERIALIZE_METHODS_PARAMS(CAddress, obj, AddrSerialization, fmt)
     {
         SER_READ(obj, obj.nTime = TIME_INIT);
-        if (fmt == AddrSerialization::DISK) {
+        switch (fmt) {
+        case AddrSerialization::DISK: {
             uint32_t disk_version = DISK_VERSION;
             READWRITE(disk_version);
             READWRITE(obj.nTime);
-        } else if (fmt == AddrSerialization::NETWORK_WITHTIME) {
+            break;
+        }
+        case AddrSerialization::NETWORK_WITHTIME: {
             READWRITE(obj.nTime);
+            break;
+        }
+        case AddrSerialization::NETWORK_NOTIME: {
+            // The only time we serialize a CAddress object without nTime is in
+            // the initial VERSION messages which contain two CAddress records.
+            break;
         }
+        } // no default case, so the compiler can warn about missing cases
         READWRITE(Using<CustomUintFormatter<8>>(obj.nServices), AsBase<CService>(obj));
     }

sipa replied:
Included.

maflcko reviewed (commented):
Concept ACK, no idea why it is not compatible with the live network

8d1b1df474db73d3010324f728cb8b7851036b4f

dongcarl commented:
Concept ACK as it will allow for a previously hacky solution to be neater. See https://github.com/bitcoin/bitcoin/issues/19477 for more details

Perhaps a worthwhile followup: Completing the example documentations of our serialization "helpers" (seems like there's already a few examples, but not all of them are documented)

sipa commented on src/protocol.h:
```diff
@@ -359,26 +359,32 @@ class CAddress : public CService
+            if (s.GetVersion() >= INIT_PROTO_VERSION) {
+                READWRITE(obj.nTime);
+            }
         }
```
There was a bug (`>= INIT_PROTO_VERSION` instead of `!= INIT_PROTO_VERSION`), fixed.

sipa commented:
Rebased after #19486 merge. Addressed comments.

sipa commented on src/serialize.h:
```diff
@@ -165,11 +165,10 @@ enum
 //! Convert the reference base type to X, without changing constness or reference type.
-template<typename X> X& ReadWriteAsHelper(X& x) { return x; }
-template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
+template<typename X> X& AsBase(X& x) { return x; }
```
It needs to maintain constness. During serialization, the object may be `const T`, while during deserialization it cannot be `const`. This function is effectively a static cast, but one that passes through constness.

sipa replied:
Added more comments to clarify this.

vasild replied:
I checked that replacing `s/AsBase/static_cast/` all over the place compiles and the few tests I ran pass.

sipa replied:
Yes, that's expected - but it may lead to UB. `static_cast` may cast away the constness of a reference, in a place where it is illegal to do so. So `AsBase` is just a more restrictive, safer, version of `static_cast`.

MarcoFalke commented:
Looks like the fuzzers don't compile on some of the commits

naumenkogs commented:
Concept ACK.

sipa commented on src/serialize.h:
```diff
@@ -1114,4 +1129,71 @@ size_t GetSerializeSizeMany(int nVersion, const T&... t)
     return sc.size();
 }
 
+/** Wrapper that overrides thet GetParams() function of a stream. */
```
Fixed.

sipa commented:
@dongcarl Agree - I've added more comments to explain the feature for now, but a more comprehensive overview of all the things the serialization framework supports would be a nice further improvement.

@MarcoFalke Hopefully fixed. I also added all variants of deserialization to the fuzzers.

vasild commented:
Code coverage report for this PR from unit + functional tests (lines not modified by this PR are dimmed and files not modified by this PR are omitted from the report):
* [report](https://people.freebsd.org/~vd/pr19503_3d3b09543_coverage/)
* [links to modified and not covered lines](https://people.freebsd.org/~vd/pr19503_3d3b09543_coverage/modified_and_not_covered.html)
* just [net_processing.cpp:3942](https://people.freebsd.org/~vd/pr19503_3d3b09543_coverage/src/net_processing.cpp.gcov.html#3942) is "interesting" (it is modified by this PR and not covered by tests).

jnewbery commented:
> just net_processing.cpp:3942 is "interesting" (it is modified by this PR and not covered by tests).

That line is pretty much a no-op and could be removed. It's ensuring that we never send more than 1000 CAddress records in a single ADDR message. However, the ADDR message is populated from vAddrToSend, which itself is limited to 1000 entries (see the PushAddress() function). So we could just remove that entire if block (and potentially replace it with an `assert(vAddr.size() <= 1000)` after the for loop.

jnewbery reviewed (commented):
concept ACK and code review ACK of the changes in protocol.h. The syntax/interface for `WithParams()` and `SERIALIZE_METHODS_PARAMS` seems reasonable, but I find it very difficult to understand the template code in serialize.h so I wasn't able to give it a thorough review.

vasild commented:
In order to check that this works as intended and to understand it better I wrote some unit tests that exercise it.

Some observations, correct me if I am wrong:

* I think it is impossible to use two separate methods `Serialize()` and `Unserialize()` with this mechanism and one has to use:
```cpp
SERIALIZE_METHODS_PARAMS(...)
{
    if (ser_action.ForRead()) {
        ... all unserialize goes here ...
    } else {
        ... all serialize goes here ...
    }
}
```

* If there are two classes `Base` and `Derived` (the second one derived from the first), both using this mechanism, then for the derived class one has to create a formatter config that specifies how to format both `Base` and `Derived`:
```cpp
Derived derived;
BaseAndDerivedFormat fmt;

fmt.m_base_format = ...
fmt.m_derived_format = ...

stream << WithParams(fmt, derived);

// and then in `Derived`:

SERIALIZE_METHODS_PARAMS(Derived, obj, DerivedAndBaseFormat, fmt)
{
    READWRITE(WithParams(fmt.m_base_format, AsBase<Base>(obj)));

    ... ser/unser this Derived object according to fmt.m_derived_format ...
}
```

<details>
<summary>unit tests, added to src/test/serialize_tests.cpp</summary>

```cpp
#include <util/strencodings.h>

#include <algorithm>
#include <ios>
#include <string>

enum class BaseFormat {
    DEC,
    HEX,
};

/// (Un)serialize a number as a string either from/to decimal or hexadecimal.
class Base
{
public:
    uint8_t m_base_data;

    Base() : m_base_data(17) {}
    explicit Base(uint8_t data) : m_base_data(data) {}

    SERIALIZE_METHODS_PARAMS(Base, obj, BaseFormat, fmt)
    {
        if (ser_action.ForRead()) {
            std::string str;
            s >> str;
            const uint8_t data = (uint8_t)std::stoul(str, 0, fmt == BaseFormat::DEC ? 10 : 16);
            SER_READ(obj, obj.m_base_data = data);
        } else {
            s << strprintf(fmt == BaseFormat::DEC ? "%02hhu" : "%02hhX", obj.m_base_data);
        }
    }
};

class DerivedAndBaseFormat
{
public:
    BaseFormat m_base_format;

    enum class DerivedFormat {
        LOWER,
        UPPER,
    } m_derived_format;
};

class Derived : public Base
{
public:
    std::string m_derived_data;

    SERIALIZE_METHODS_PARAMS(Derived, obj, DerivedAndBaseFormat, fmt)
    {
        READWRITE(WithParams(fmt.m_base_format, AsBase<Base>(obj)));

        if (ser_action.ForRead()) {
            std::string str;
            s >> str;
            SER_READ(obj, obj.m_derived_data = str);
        } else {
            s << (fmt.m_derived_format == DerivedAndBaseFormat::DerivedFormat::LOWER ?
                      ToLower(obj.m_derived_data) :
                      ToUpper(obj.m_derived_data));
        }
    }
};

BOOST_AUTO_TEST_CASE(with_params_base)
{
    Base b{15};

    CDataStream stream(SER_DISK, PROTOCOL_VERSION);

    stream << WithParams(BaseFormat::DEC, b);
    BOOST_CHECK_EQUAL(stream.str(), "\2" "15");

    b.m_base_data = 0;
    stream >> WithParams(BaseFormat::DEC, b);
    BOOST_CHECK_EQUAL(b.m_base_data, 15);

    stream.clear();

    stream << WithParams(BaseFormat::HEX, b);
    BOOST_CHECK_EQUAL(stream.str(), "\2" "0F");

    b.m_base_data = 0;
    stream >> WithParams(BaseFormat::HEX, b);
    BOOST_CHECK_EQUAL(b.m_base_data, 0x0F);
}

BOOST_AUTO_TEST_CASE(with_params_vector_of_base)
{
    std::vector<Base> v{Base{15}, Base{255}};

    CDataStream stream(SER_DISK, PROTOCOL_VERSION);

    stream << WithParams(BaseFormat::DEC, v);
    BOOST_CHECK_EQUAL(stream.str(), "\2\2" "15" "\3" "255");

    v[0].m_base_data = 0;
    v[1].m_base_data = 0;
    stream >> WithParams(BaseFormat::DEC, v);
    BOOST_CHECK_EQUAL(v[0].m_base_data, 15);
    BOOST_CHECK_EQUAL(v[1].m_base_data, 255);

    stream.clear();

    stream << WithParams(BaseFormat::HEX, v);
    BOOST_CHECK_EQUAL(stream.str(), "\2\2" "0F" "\2" "FF");

    v[0].m_base_data = 0;
    v[1].m_base_data = 0;
    stream >> WithParams(BaseFormat::HEX, v);
    BOOST_CHECK_EQUAL(v[0].m_base_data, 0x0F);
    BOOST_CHECK_EQUAL(v[1].m_base_data, 0xFF);
}

BOOST_AUTO_TEST_CASE(with_params_derived)
{
    Derived d;
    d.m_base_data = 15;
    d.m_derived_data = "xY";

    DerivedAndBaseFormat fmt;

    CDataStream stream(SER_DISK, PROTOCOL_VERSION);

    fmt.m_base_format = BaseFormat::DEC;
    fmt.m_derived_format = DerivedAndBaseFormat::DerivedFormat::LOWER;
    stream << WithParams(fmt, d);

    fmt.m_base_format = BaseFormat::HEX;
    fmt.m_derived_format = DerivedAndBaseFormat::DerivedFormat::UPPER;
    stream << WithParams(fmt, d);

    BOOST_CHECK_EQUAL(stream.str(), "\2" "15" "\2" "xy" "\2" "0F" "\2" "XY");
}
```
</details>

jnewbery commented on src/serialize.h:
```diff
@@ -1114,4 +1178,74 @@ size_t GetSerializeSizeMany(int nVersion, const T&... t)
+public:
+    explicit ParamsWrapper(const Params& params, T obj) : m_params(params), m_object(obj) {}
+
+    //! Serialize to another ParamsStream: optimize by skipping it.
```
This may be because I don't have a good understanding of serialize.h in general so perhaps you can just ignore this, but I don't understand what's being skipped here.

sipa replied:
It's an attempt at avoiding crazy template expansion in the compiler when multiple levels of serializers each use `WithParams` (the Stream object that would hit the bottom would be ParamsStream<ParamsInner, ParamsStream<ParamsMid, ParamsStream<ParamsOuter, BaseStream>>>, and all its methods need instantiating). This optimization will just result in `ParamsStream<ParamsInner, BaseStream>`.

jnewbery commented on src/protocol.h:
```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
+            READWRITE(obj.nTime);
+            break;
+        case AddrSerialization::NETWORK_NOTIME:
             // The only time we serialize a CAddress object without nTime is in
```
Thinking about it a bit more, I actually think this whole comment can be removed. It made sense when the serialization was based on a protocol version that we don't support anywhere else (and was added to document something that caused confusion for reviewers: https://github.com/bitcoin/bitcoin/pull/14033#issuecomment-652456443), but when the serialization is based on a parameter that we pass in explicitly, it's easy enough to grep for that parameter to see where it's used.

sipa replied:
Done.

MarcoFalke commented on src/protocol.h:
```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
-    SERIALIZE_METHODS(CAddress, obj)
+    SERIALIZE_METHODS_PARAMS(CAddress, obj, AddrSerialization, fmt)
     {
         SER_READ(obj, obj.nTime = TIME_INIT);
```
This has been added here on purpose: https://github.com/bitcoin/bitcoin/pull/19020#discussion_r427620111

jnewbery replied:
The linked comment doesn't make much sense to me. There are plenty of examples where member fields aren't [de]serialized, eg:

https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/addrman.h#L48-L55
https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/primitives/block.h#L68-L69
https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/wallet/walletutil.h#L99

or only [de]serialized conditionally:

https://github.com/bitcoin/bitcoin/blob/0f6900e780afb9c176cf7d1da4a1fc0ca6c9118a/src/wallet/walletdb.h#L106-L108

vasild replied:
Could it happen that we unserialize into the same object twice? E.g.
```
s >> addr;
... do something with addr, possibly change nTime ...
s >> addr;
// if we don't explicitly set nTime during unserialization,
// then here we may end up with some leftover nTime?
```

sipa replied:
Yes, that's the reason to prefer fully initializing all fields when deserializing (or at least, all fields that are subject to serialization in the first place).

sipa replied:
@jnewbery I think in general we should aim to support deserializing into an already existing (and not just default-initialized) objects, as it's hard to enforce the opposite.

`CAddrInfo` is a bit of a special case, as its memory-only fields are "managed" by CAddrMan more than being a standalone object. In the other cases, I think it makes sense to make deserialization reset those fields.

jnewbery replied:
Thanks for the additional background. My immediate reaction is that it seems dangerous to re-deserialize into an existing object and that it could lead to subtle, confusing bugs. However, we've definitely wondered off-topic for this PR, so I'm going to mark this conversation as resolved. If the expectation really is that the deserialization methods overwrite all fields, a future PR could update the documentation for `SERIALIZE_METHODS` to explicitly state that.

MarcoFalke commented on src/protocol.h:
```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
+    /** The disk serialization for CAddress includes a version number.
+     *  Traditionally the number CLIENT_VERSION was used, but it has since
+     *  been disentangled from it. */
+    static constexpr uint32_t DISK_VERSION{210000};
```
It is the CLIENT_VERSION of the next major release. The number should be equal to or greater than the largest CLIENT_VERSION of any released version.

MarcoFalke commented:
Will review when travis is green

sipa committed:
Add deserialization fuzzers for address variants

sipa commented:
@MarcoFalke I've been pushing fixes, and GitHub's UI doesn't seem to be picking them up...

sipa commented:
https://github.com/sipa/bitcoin/tree/202007_ser_options has commits that aren't here.

MarcoFalke commented:
No idea what's going on. Since this has no reviews, you could try rebasing once more on master and force push? :man_shrugging:

sipa commented:
Ah it was just behind; it picked them up now.

MarcoFalke commented on src/serialize.h:
```diff
@@ -1249,3 +1249,4 @@ template<typename Params, typename T>
 static inline ParamsWrapper<Params, T&> WithParams(const Params& params, T&& t) { return ParamsWrapper<Params, T&>(params, t); }
 
 #endif // BITCOIN_SERIALIZE_H
+
```
Can be removed now

sipa replied:
Done.

MarcoFalke commented:
Travis still failing in the fuzzer. Also, needs rebase after https://github.com/bitcoin/bitcoin/pull/19360#discussion_r455193993

jnewbery commented:
I think this requires a change to netaddress.h now that #19360 has been merged: https://github.com/bitcoin/bitcoin/pull/19360/files#diff-76d15e11a95be7a4aee1eb89de6098caR165

vasild committed:
test: add tests that exercise WithParams()

vasild commented:
Alright, so this:

```cpp
    SERIALIZE_METHODS_PARAMS(Base, obj, BaseFormat, fmt)
    {
        if (ser_action.ForRead()) {
            std::string str;
            s >> str;
            const int base = fmt == BaseFormat::DEC ? 10 : 16;
            const uint8_t data = (uint8_t)std::stoul(str, 0, base);
            SER_READ(obj, obj.m_base_data = data);
        } else {
            s << strprintf(fmt == BaseFormat::DEC ? "%02hhu" : "%02hhX", obj.m_base_data);
        }
    }
```

is equivalent to this:

```cpp
    template <typename Stream>
    void Serialize(Stream& s) const
    {
        s << strprintf(s.GetParams() == BaseFormat::DEC ? "%02hhu" : "%02hhX", m_base_data);
    }

    template <typename Stream>
    void Unserialize(Stream& s)
    {
        std::string str;
        s >> str;
        const int base = s.GetParams() == BaseFormat::DEC ? 10 : 16;
        m_base_data = (uint8_t)std::stoul(str, 0, base);
    }
```

I pushed the unit tests that exercise this feature at https://github.com/vasild/bitcoin/commit/8d89c1d43, feel free to pick (some of) them into this PR if you find them interesting, or if they may serve as an example of how to use this.

vasild commented on src/protocol.h:
```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
     return (services & NODE_NETWORK) || (services & NODE_NETWORK_LIMITED);
 }
 
+enum class AddrSerialization {
```
nit, feel free to ignore: since this is also used in **un**serialization, the name could be misleading. `AddrFormat`?

sipa replied:
I've changed it to `CAddress::Format`.

vasild reviewed (approved):
ACK fdf66d325

Just some nits below and another nit at https://github.com/bitcoin/bitcoin/pull/19503#discussion_r454942622

sipa commented on src/net_processing.cpp:
```diff
@@ -477,7 +477,7 @@ static void PushNodeVersion(CNode& pnode, CConnman* connman, int64_t nTime)
     CAddress addrMe = CAddress(CService(), nLocalNodeServices);
 
-    connman->PushMessage(&pnode, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::VERSION, PROTOCOL_VERSION, (uint64_t)nLocalNodeServices, nTime, addrYou, addrMe,
+    connman->PushMessage(&pnode, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::VERSION, PROTOCOL_VERSION, (uint64_t)nLocalNodeServices, nTime, WithParams(AddrSerialization::NETWORK_NOTIME, addrYou), WithParams(AddrSerialization::NETWORK_NOTIME, addrMe),
```
Yeah, it was ridiculously long. I've split it up.

Indeed, `INIT_PROTO_VERSION` is no longer needed here, but I think it's better to just get rid of all of these version numbers once everything has been updated to not using them anymore.

sipa commented on src/serialize.h:
```diff
@@ -197,14 +214,43 @@ template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
+ * stack; it is passed down recursively into all serialization code, until
+ * another WithParams overrides it.
+ *
+ * Compilation will fail in any context where serialization is invoke but
```
Done.

sipa commented:
Rebased, added a few more comments, renamed `AddrSerialization` to `CAddress::Format`, and included @vasild's unit test.

sipa commented on src/protocol.h:
```diff
@@ -354,34 +354,45 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
-        int nVersion = s.GetVersion();
-        if (s.GetType() & SER_DISK) {
-            READWRITE(nVersion);
+        switch (fmt) {
```
I agree. I've reverted it to the if/then/else style.

sipa commented on src/serialize.h:
```diff
@@ -197,14 +214,43 @@ template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
- * and non-const for deserializing, without casts.
+/** Variant of FORMATTER_METHODS that supports a declared parameter type.
+ *
+ * With a formatter has a declared parameter type, it must be invoked directly or
```
"If" is even better, I think? Done.

vasild reviewed (approved):
ACK 483ed4574

Thanks for adding input validation to the hex/dec examples in the unit tests!

jonatack reviewed (commented):
ACK 483ed45 code review over several sessions this week, verified the build at each commit, ran the 3 address deserialization fuzz tests as a sanity check. Thanks for the unit test coverage -- a good addition and demonstrates use of the feature. Edit: am running a node with this branch.

ryanofsky commented on src/test/serialize_tests.cpp:
```diff
@@ -376,4 +377,175 @@ BOOST_AUTO_TEST_CASE(class_methods)
+            uint32_t data;
+            bool ok;
+            if (fmt == BaseFormat::DEC) {
+                ok = ParseUint32(str, &data);
```
In commit "test: add tests that exercise WithParams()" (483ed457411d164ad71115dbfe1d2b4ff8b2e5f3)

Think this needs to capitalize Int: `ParseUInt32`

ryanofsky reviewed (approved):
Code review ACK 483ed457411d164ad71115dbfe1d2b4ff8b2e5f3.

This seems all right, but I think it's significantly more complex than it needs to because it is making serialization with params something that can be done in conjunction with using a formatter, instead of just a special case of formatting. Simplification along lines of 106d8039b61ff4b771dc996e3ab9156019ba4470 [branch](http://github.com/ryanofsky/bitcoin/commits/review.19503.2-edit)) seems to work (only ran tests locally).

sipa commented:
@ryanofsky That approach works and is a lot simpler, but I think it misses the "automatic recursion" property I was hoping for, that you could e.g. serialize a CBlock with specified CTransaction parameters, and it would transparently pass those parameters down. Or a block could define its own type that can be converted to the CTransaction parameters type, but not need to anything beyond that.

Whether that's a desirable depends how much you treat a higher-level structure

MarcoFalke commented:
re-ACK 483ed457411d164ad71115dbfe1d2b4ff8b2e5f3, though I did not review the serialize.h changes 🚤

Changes since last review:
* Rebase
* Added comments/doc
* Fix "off-by-one" issue/typo bug in commit "12a5a3c032 Disentangle disk address version from client version"
* Properly deserialize addrFrom without time (Bugfix in commit "21446c5b5d Use serialization parameters for CAddress serialization")
* Fix and extend fuzzers
* Add tests from vasild

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 483ed457411d164ad71115dbfe1d2b4ff8b2e5f3, though I did not review the serialize.h changes 🚤

Changes since last review:
* Rebase
* Added comments/doc
* Fix "off-by-one" issue/typo bug in commit "12a5a3c032 Disentangle disk address version from client version"
* Properly deserialize addrFrom without time (Bugfix in commit "21446c5b5d Use serialization parameters for CAddress serialization")
* Fix and extend fuzzers
* Add tests from vasild
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiZTQwAoa37uqI1WTtHnkBHX/T8VKa36Ba6GzwdKgCSLVl7uMr0qw5jL/UrHuG+
tHm4W6GH6fuxYSblm+r7ujigrAkD4/3gPsbOG5mHaYTDOtiL9/ZTHT4HT/nbyZBL
KmaVaLHihfZy8NS05IG8y4sB97ZWXmiAak4An6NdfmjtB5ahWVH5HuLxLERIM7nb
yLjW/5GVbnnMvhirK/nyMcAAXSlTY5twmjT2LYWQ0TnJnNF0djY0VMkvNF8RrvgZ
k3zVnE0J+MW5mv2KcKRCYouGXc+WK777s5gc7bVOfP0J0So9i20LP1pFGqdR+FhV
0WF0aO27EJVePxev36zdaF4CuEFadP99NN5fKnY6rUwpd7+3hIRWZOH2958bbaSH
3oE+M5tLKd+qsAIZ+Lz7W3XU/oK9DE3v8YRasuLzgAFXoiZVkYnISWp/N+FZv+/6
rhAFSyF45HGFkEAa40KdzPV/aT+LWi9UnZe+wa1RtOoiM25jDMTbyLBJiJNz3EoT
vjuegm6r
=lpi2
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c6299c40a36d62d1a32997018080b37d4feeef88382e17059f749fec43043e26  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c6299c40a36d62d1a32997018080b37d4feeef88382e17059f749fec43043e26f010f238cb3992877e848944f76a0343dd0908fff0102aa566b34d009b04762788277f6ac9d608f1045f295cb2f008e11f8e3009ba9b280083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103ddf7cadfbd958de14b17793de8b175d08f1045f295caef0081101c1afe0c3fc390083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010b20b5d0687ab4e4e734568c9f9f70e6a08f1045f295caff008b48263468e1d50b10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
Note to myself, if this gets merged: It would be good to remove the now unused protocol version includes. I.e.


```diff
commit 7ffff5b3237fa418637d30ddb59534ab47d3f610 (HEAD)
Author: MarcoFalke <falke.marco@gmail.com>
Date:   Tue Aug 4 14:58:25 2020 +0200

    Remove unused protocol version header from protocol.h

diff --git a/src/bench/rpc_blockchain.cpp b/src/bench/rpc_blockchain.cpp
index 4b45264a3c..09bfcb60a5 100644
--- a/src/bench/rpc_blockchain.cpp
+++ b/src/bench/rpc_blockchain.cpp
@@ -8,6 +8,7 @@
 #include <rpc/blockchain.h>
 #include <streams.h>
 #include <validation.h>
+#include <version.h> // For PROTOCOL_VERSION
 
 #include <univalue.h>
 
diff --git a/src/protocol.h b/src/protocol.h
index a564508f89..58669e20b5 100644
--- a/src/protocol.h
+++ b/src/protocol.h
@@ -14,7 +14,6 @@
 #include <primitives/transaction.h>
 #include <serialize.h>
 #include <uint256.h>
-#include <version.h>
 
 #include <stdint.h>
 #include <string>
```

and

```diff
commit 4550d37e268e7ffe34b3f5ea11d6b3f231d9065a
Author: MarcoFalke <falke.marco@gmail.com>
Date:   Tue Aug 4 12:56:55 2020 +0200

    Remove unused protocol version header from undo.h

diff --git a/src/undo.h b/src/undo.h
index a98f046735..1fb9ac0688 100644
--- a/src/undo.h
+++ b/src/undo.h
@@ -11,7 +11,6 @@
 #include <consensus/consensus.h>
 #include <primitives/transaction.h>
 #include <serialize.h>
-#include <version.h>
 
 /** Formatter for undo information for a CTxIn
  *

MarcoFalke commented on src/protocol.h:
```diff
@@ -354,34 +354,38 @@ static inline bool MayHaveUsefulAddressDB(ServiceFlags services)
+    SERIALIZE_METHODS_PARAMS(CAddress, obj, Format, fmt)
     {
         SER_READ(obj, obj.nTime = TIME_INIT);
-        int nVersion = s.GetVersion();
```
One include can now be removed, see https://github.com/bitcoin/bitcoin/pull/19503#issuecomment-668585730

maflcko reviewed (commented):
Still needs rebase (trivial)

jonatack commented:
Needs rebase.

vasild commented:
This would resolve https://github.com/bitcoin/bitcoin/issues/19477.

Also, `ADDRV2_FORMAT` could be replaced with the technique introduced in this PR.

MarcoFalke commented:
Could also use `LIFETIMEBOUND` as per https://github.com/bitcoin/bitcoin/pull/19387#issuecomment-658372256?

MarcoFalke commented:
@sipa Let us know if someone should pick this up. I might take a stab at it.

sipa commented:
@MarcoFalke Not going to work on this for the time being.

fanquake commented:
Removing up for grabs as this has mostly been picked up. i.e #25284.
